### PR TITLE
Update Rails to 6.0 

### PR DESCRIPTION
**Pull request by thalesmiguel**
**What this PR does / why we need it**:

This PR attempts to update Porta's Rails version to 6.0.x. It also tries to update the minimum amount of gems necessary for the Rails update.

**Which issue(s) this PR fixes** 

[THREESCALE-7405](https://issues.redhat.com/browse/THREESCALE-7405)

**Verification steps** 

That's a good question. Put it on `staging` and/or the development cluster and try to break it? 

**Special notes for your reviewer**:

### CODE REVIEW

File: app/forms/notification_preferences_form.rb [hunk 1]
```diff
@@ -4,6 +4,7 @@ class NotificationPreferencesForm < Reform::Form
   attr_reader :current_user, :user_ability
 
   delegate :update_attributes,    to: :model
+  delegate :update,               to: :model
```
Comment (mayorova) [inline app/forms/notification_preferences_form.rb:7]: Probably line 6 (`delegate :update_attributes`) can be deleted now?

File: app/lib/backend/errors.rb [hunk 2]
```diff
@@ -13,7 +13,7 @@ module Backend
     'provider.invalid_transaction_id' => 'transaction does not exist'
   }.freeze
 
-  class Error < StandardError
+  class Errors < StandardError
```
Comment (mayorova) [inline app/lib/backend/errors.rb:16]: This class name is weird, I think I'd prefer to rename the file.

File: app/models/settings.rb [hunk 3]
```diff
@@ -44,6 +44,7 @@ def update_attributes(attributes)
 
     super(attributes)
   end
+  alias update_attributes update
```
Comment (mayorova) [inline app/models/settings.rb:47]: Do we need this? Can we just remove `update_attributes` usage?

File: app/models/web_hook/event.rb [hunk 4]
```diff
@@ -39,7 +39,7 @@ def model
 
     delegate :logger, to: :Rails
 
-    def has_transactional_callbacks?
+    def trigger_transactional_callbacks?
```
Comment (mayorova) [inline app/models/web_hook/event.rb:42]: What's the reason for this change?...

File: config/application.rb [hunk 5]
```diff
@@ -34,29 +60,26 @@ class Application < Rails::Application
...
     config.active_record.include_root_in_json = true
     # Make `form_with` generate non-remote forms. Defaults true in Rails 5.1 to 6.0
+    config.action_view.form_with_generates_remote_forms = false
+    # Make Ruby preserve the timezone of the receiver when calling `to_time`.
+    config.active_support.to_time_preserves_timezone = false
+    config.action_dispatch.use_authenticated_cookie_encryption = false
+    config.active_support.use_authenticated_message_encryption = false
+
+    # Use a modern approved hashing function
+    config.active_support.hash_digest_class = OpenSSL::Digest::SHA256
```
Comment (mayorova) [inline config/application.rb:74]: We can leave a note that this is the default in Rails 7.0, so can be removed when we upgrade.

File: config/application.rb [hunk 6]
```diff
@@ -263,12 +286,7 @@ def try_config_for(*args)
       config.three_scale.oauth2 = config_for(:oauth2)
     end
 
-    initializer :log_formatter, after: :initialize_logger do
-      config.log_formatter = System::ErrorReporting::LogFormatter.new
-      (config.logger || Rails.logger).formatter = config.log_formatter
-    end
-
-    config.paperclip_defaults = {
+    Rails.application.config.paperclip_defaults = {
```
Comment (mayorova) [inline config/application.rb:289]: just wondering why Paperclip is different, and the config is accessed through `Rails.application`

File: config/initializers/new_framework_defaults_6_0.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
+# Be sure to restart your server when you modify this file.
```
Comment (mayorova) [inline config/initializers/new_framework_defaults_6_0.rb:1]: Now that `config.load_defaults` is set to `6.0` I guess this file can be deleted too?

File: app/forms/notification_preferences_form.rb (hunk 1 above)
Comment (jlledom) [reply app/forms/notification_preferences_form.rb:7]: I think so. Done.

File: app/lib/backend/errors.rb (hunk 2 above)
Comment (jlledom) [reply app/lib/backend/errors.rb:16]: True. Done.

File: app/models/settings.rb (hunk 3 above)
Comment (jlledom) [reply app/models/settings.rb:47]: I don't know why I added this line. I removed it, let's see if something fails.

File: app/lib/finance/api/response.rb [hunk 8]
```diff
@@ -1,6 +1,6 @@
 module Finance
-  module API
-    module Collection
+  module Api
+    module Response
```
[the remark this answers is not in the document]
Comment (jlledom) [reply app/lib/finance/api/response.rb:3]: I don't know, it seems a non-sense for me. The file was called `response.rb` but the class was called `Collection`, so this wasn't being loaded by the autoloader AFAIK. Besides, there is a `collection.rb` file on the same [folder](https://github.com/3scale/porta/tree/master/app/lib/finance/api) which is actually defining `Collection`. Nothing of this seems to make any sense so I'm deleting this file a let's see if any test fails.

File: app/models/cinstance.rb [hunk 9]
```diff
@@ -515,5 +515,3 @@ def generate_key
     plan.issuer.prefix_key(SecureRandom.hex(16))
   end
 end
-
-ApplicationContract = Cinstance
```
[the remark this answers is not in the document]
Comment (jlledom) [reply app/models/cinstance.rb:519]: When I removed this alias, I searched for `ApplicationContract` in the code and only found one reference. See the [commit](https://github.com/3scale/porta/pull/3339/commits/dc0c6c39d283499f56d2f65b881db122faa52741). There could always be something like that cucumber step you pointed out, which generates the class name from an interpolated string:

https://github.com/3scale/porta/blob/8bcc7f77bf776619d2eea4a4bc1b1e6b0abfaec9/features/step_definitions/contracts/contract_steps.rb#L9

I'm not sure that test is right, though, because the `type` parameter is transformed by this:

https://github.com/3scale/porta/blob/5c6bbdd999accca358bf16566abe56fdd0ee0c92/features/support/parameter_types.rb#L96-L100

And the resulting `type` can only be `"account_plan"`, `"service_plan"` or `"application_plan"`, which generate class names like `AccountPlanContract`, `ServicePlanContract` and `ApplicationPlanContract`, which don't exist. I think the cucumbers [using](https://github.com/3scale/porta/blob/f6e98d1d8a66f28d13a91ebba8767e12814509ae/features/old/signup/buyers.feature#L39) this step should fail, but they don't, so I assume we can't trust the tests on this. I created a new `ApplicationContract` class in a way it's properly loaded by the autoloader.

[the remark this answers is not in the document]
Comment (mayorova) [reply app/models/cinstance.rb:519]: Hm, but I think in [this step](https://github.com/3scale/porta/blob/8bcc7f77bf776619d2eea4a4bc1b1e6b0abfaec9/features/step_definitions/contracts/contract_steps.rb#L3) the parameter type `{plan_type}` is not used. This step uses a simple regex, so `type` would be either of these values: `account|service|application`. This step is used in this test, I believe: https://github.com/3scale/porta/blob/8bcc7f77bf776619d2eea4a4bc1b1e6b0abfaec9/features/old/signup/buyers.feature#L37-L39

On the other hand, the `plan_type` parameter is used in this step:
https://github.com/3scale/porta/blob/8bcc7f77bf776619d2eea4a4bc1b1e6b0abfaec9/features/step_definitions/plans_steps.rb#L5-L9

I think this is the example step: https://github.com/3scale/porta/blob/8bcc7f77bf776619d2eea4a4bc1b1e6b0abfaec9/features/old/applications/providers/search.feature#L12-L15

So, are you saying that removing `ApplicationContract` alias [this test](https://github.com/3scale/porta/blob/f6e98d1d8a66f28d13a91ebba8767e12814509ae/features/old/signup/buyers.feature#L39) was not failing?...

File: app/models/web_hook/event.rb (hunk 4 above)
Comment (jlledom) [reply app/models/web_hook/event.rb:42]: I've been reading a bit and it seems to be required to run callbacks after a DB transaction succeeded or rollbacked.

In rails 6, `trigger_transactional_callbacks?` is expected to be true in order to actually call the callbacks. See [#36190](https://github.com/rails/rails/pull/36190/files#diff-dd0ea9dfea1fd486f2b1b5eac8c798f839467173a1c2315d3ad04f4098f7dda3).

The two files affected by this change are [/app/models/web_hook/event.rb](https://github.com/3scale/porta/blob/00d75d0f5e9a8cb936632243ca330c094488a0e6/app/models/web_hook/event.rb) and [/app/subscribers/after_commit_subscriber.rb](https://github.com/3scale/porta/blob/632667de4bade6a0cf2859341499d9a9363d68a9/app/subscribers/after_commit_subscriber.rb). You can compare the structure of these files with the one showed at this article:

https://www.dmitry-ishkov.com/2019/04/running-activejob-after-transaction.html

I don't really know how to verify this behavior, other than running the tests for the affected files: [/test/unit/web_hook/event_test.rb](https://github.com/3scale/porta/blob/e8feceb1e8950912b88d1273d3b1e5aeb4ac5fc1/test/unit/web_hook/event_test.rb) and [/test/subscribers/after_commit_subscriber_test.rb](https://github.com/3scale/porta/blob/1a48b303b61f4893e7eaa002dd4cf73f4474ab6d/test/subscribers/after_commit_subscriber_test.rb)

If we run these tests without the change, they fail with this error:

`"Minitest::UnexpectedError: NoMethodError: undefined method 'trigger_transactional_callbacks?'"`

After the change, the tests pass. I think we can assume this is correct, what do you think?

File: app/models/cinstance.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (jlledom) [reply app/models/cinstance.rb:519]: We probably found a bug. I checked it with the debugger, the step receives `type='application_plan'`, when it should receive `type='application'`. Following the stack back to the place where the value is actually set, [here](https://github.com/cucumber/cucumber-expressions/blob/main/ruby/lib/cucumber/cucumber_expressions/argument.rb#L33), `@parameter_type.name` is `'plan_type'`. So the mentioned transformer is being invoked.

> So, are you saying that removing ApplicationContract alias [this test](https://github.com/3scale/porta/blob/f6e98d1d8a66f28d13a91ebba8767e12814509ae/features/old/signup/buyers.feature#L39) was not failing?...

Exactly.

File: config/application.rb (hunk 6 above)
Comment (jlledom) [reply config/application.rb:290]: Because I moved this inside a `config.to_prepare` block, where `config` is out of scope, to fix the autoloader. Later, I found a problem and had to move it out again([commit](https://github.com/3scale/porta/commit/92f6bbffa9f72d7b6cad4c8ecdc00eae09930b4b)). And forgot to remove the `Rails.application` part. Fixed.

File: config/initializers/new_framework_defaults_6_0.rb (hunk 7 above)
Comment (jlledom) [reply config/initializers/new_framework_defaults_6_0.rb]: It was actually setting a couple of parameters. I moved them to `application.rb` and deleted the file.

File: config/application.rb (hunk 6 above)
Comment (mayorova) [reply config/application.rb]: Speaking of which... Why some many initializers are wrapped into `config.to_prepare` in the first place? Is that required?

From reading about this method, it seems that it's to ensure that the config is reloaded on every request (for dev). But again - do we need this?... :thinking: 

Was this to silence the warnings? I've seen initializers may cause warnings like:
```
DEPRECATION WARNING: Initialization autoloaded the constant XXX
```

Comment (jlledom) [reply config/application.rb]: `config.to_prepare` silences that deprecation warning, indeed. And it's required for two reasons:

1. Without it, we'd could experience weird errors at any moment after making changes to `*.rb` files in our dev environments. In practice, we would need to restart the server after every change to avoid the errors.
2. In rails 7 the deprecation warning has been deleted and an error is returned instead. So in our case porta wouldn't even start in rails 7.

The blocks under `config.to_prepare` are not called for every request, but for every Zeitwerk reload. That only happens when a request is received **and** there are changes in any `.rb` file. So only the first request after code changed triggers the reload.

Comment (mayorova) [reply config/application.rb]: Thanks for the explanation!

File: app/models/web_hook/event.rb (hunk 4 above)
Comment (mayorova) [reply app/models/web_hook/event.rb:42]: I looked into it, and while I still do not understand completely these ActiveRecord internals, I think it makes sense.
The code is trying to make the classes that are not AR classes act like AR classes, e.g. only do things when the transaction was committed, and not otherwise. Like do not send a web hook if the transaction rolled back for some reason.

**Comment by mayorova** [comment]
I was trying to test the web hooks feature, and noticed that multiple Sidekiq workers fail with the following error:
```
NameError: uninitialized constant ThreeScale::Analytics::SidekiqMiddleware::UserTracking Did you mean? ThreeScale::Analytics::UserTracking
```

It seems to be affecting any worker, the easiest way to test probably is to make any change on an audited model, to trigger a AuditedWorker

**Comment by jlledom** [comment]
> I was trying to test the web hooks feature, and noticed that multiple Sidekiq workers fail with the following error:
> 
> ```
> NameError: uninitialized constant ThreeScale::Analytics::SidekiqMiddleware::UserTracking Did you mean? ThreeScale::Analytics::UserTracking
> ```

Good catch. Fixed!

File: config/cable.yml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+development:
```
Comment (mayorova) [inline config/cable.yml:1]: If this file is required, we need to add one in `openshift/system/config`

But I'd say we don't need to enable it (i.e. don't use redis)

Comment (akostadinov) [reply config/cable.yml:1]: Maybe rails automatic update added it. We better get rid of it.

Comment (jlledom) [reply config/cable.yml:1]: As far as I know, that file is for configuring `ActionCable`, which we removed because it's not being used. See: https://github.com/3scale/porta/pull/3339/files#diff-c1fd91cb1911a0512578b99f657554526f3e1421decdb9e908712beab57e10f9R12-R29

I'm removing it.

Comment (mayorova) [reply config/cable.yml]: Oh even better!
I removed it previously from the config examples https://github.com/3scale/porta/pull/3225/commits/a4947ffeabfe3befa8fbc06c07442108a23433bc#diff-0c400262ffebf26a5f82d65ac15697731ec1917292851cda09a9687ed8bd1559

But I thought that maybe it's required now in Rails 6. Great that we just disable action cable :+1:

**Comment by jlledom** [comment]
Closing, as it's been superseded by https://github.com/3scale/porta/pull/3456

**Comment by jlledom** [comment]
Closing, as it's been superseded by https://github.com/3scale/porta/pull/3456
